File: packages/ts/src/structural.ts
// Structural pass: REUSES (does not reimplement) the official JSON-stat 2.0 JSON Schema 2020-12
// definitions vendored in schemas/vendored/. Violations are normalized into the same Finding shape
// the semantic engine uses, with code STRUCTURAL_VIOLATION and the ajv keyword retained in `meta`.
import { readFileSync } from "node:fs";
import { fileURLToPath } from "node:url";
import path from "node:path";
import type { Finding } from "./types.js";
import { lookupRule } from "./manifest.js";

// ajv ships the 2020-12 dialect as a separate build; its bundled types resolve at this path.
// @ts-ignore - ajv/dist/2020.js declaration resolution varies across ajv versions
import Ajv2020 from "ajv/dist/2020.js";
import addFormats from "ajv-formats";

const here = path.dirname(fileURLToPath(import.meta.url));
const vendoredDir = path.resolve(here, "../../../schemas/vendored");

function loadSchema(name: string): any {
  return JSON.parse(readFileSync(path.join(vendoredDir, name), "utf8"));
}

// The vendored schemas use `\-` (escaped hyphen) in the `updated` date pattern, which is an
// invalid escape under a Unicode-flag RegExp. None of the schemas use Unicode property escapes,
// so disabling unicodeRegExp keeps identical matching while letting the patterns compile.
const ajv = new (Ajv2020 as any)({ allErrors: true, strict: false, unicodeRegExp: false });
(addFormats as any)(ajv);

const compiled = {
  dataset: ajv.compile(loadSchema("dataset.json")),
  collection: ajv.compile(loadSchema("collection.json")),
  dimension: ajv.compile(loadSchema("dimension.json")),
  index: ajv.compile(loadSchema("index.json")),
};

function pickValidator(doc: any): (data: any) => boolean {
  const cls = doc?.class;
  if (cls === "dataset") return compiled.dataset;
  if (cls === "collection") return compiled.collection;
  if (cls === "dimension") return compiled.dimension;
  return compiled.index; // general oneOf of all three classes
}

export function validateStructural(doc: any): { ok: boolean; findings: Finding[] } {
  const validate = pickValidator(doc);
  if (validate(doc)) return { ok: true, findings: [] };
  const r = lookupRule("STRUCT");
  const errs = (validate as any).errors ?? [];
  const findings: Finding[] = errs.map((e: any) => ({
    code: r.code,
    ruleId: r.id,
    severity: r.severity,
    specRef: r.specRef,
    path: e.instancePath || "/",
    message: e.message ?? "JSON Schema violation",
    meta: { keyword: e.keyword, schemaPath: e.schemaPath, params: e.params },
  }));
  return { ok: false, findings };
}

File: crates/validator/src/structural.rs
// Structural pass: REUSES the official JSON-stat 2.0 JSON Schema 2020-12 definitions, embedded at
// compile time from schemas/vendored/. Violations are normalized into the Finding shape with
// code STRUCTURAL_VIOLATION.
//
// In jsonschema 0.22, Validator::validate returns Err(Box<dyn Iterator<Item = ValidationError>>),
// i.e. ALL violations, so we map over them.
use crate::manifest::{lookup, ManifestData};
use crate::types::{Finding, Severity};
use jsonschema::validator_for;
use serde_json::Value;

// Embedded from the copies synced into src/_vendored/schemas/ by build.rs, which mirrors the
// monorepo-root ../../schemas/vendored/*.json. Using local copies keeps the crate self-contained for
// `cargo publish` (which only packs files inside the package directory).
const DATASET_SCHEMA: &str = include_str!("_vendored/schemas/dataset.json");
const COLLECTION_SCHEMA: &str = include_str!("_vendored/schemas/collection.json");
const DIMENSION_SCHEMA: &str = include_str!("_vendored/schemas/dimension.json");
const INDEX_SCHEMA: &str = include_str!("_vendored/schemas/index.json");

fn pick_schema_str(doc: &Value) -> &'static str {
    match doc["class"].as_str() {
        Some("dataset") => DATASET_SCHEMA,
        Some("collection") => COLLECTION_SCHEMA,
        Some("dimension") => DIMENSION_SCHEMA,
        _ => INDEX_SCHEMA, // general oneOf of all three classes
    }
}

pub fn validate_structural(doc: &Value, mf: &ManifestData) -> Vec<Finding> {
    let schema: Value = serde_json::from_str(pick_schema_str(doc))
        .expect("embedded curated schema must be valid JSON");
    // As of 1.0.0 the structural pass loads the curated/bundled schemas, which write the `updated`
    // date hyphen literally and compile cleanly (the vendored upstream originals' invalid `\-`
    // escape that previously forced a silent-skip here is gone). A compile failure now indicates a
    // real schema regression, so propagate it as a panic rather than silently no-op-ing.
    let validator = validator_for(&schema)
        .expect("embedded curated schema must compile under the jsonschema engine");
    let r = lookup(mf, "STRUCT");
    // Bind to a local so the validate() temporary (which borrows `validator`) is dropped before
    // `validator` itself, satisfying the borrow checker.
    let findings = match validator.validate(doc) {
        Ok(()) => Vec::new(),
        Err(errs) => errs
            .map(|e| {
                let raw = e.instance_path.to_string();
                let path = if raw.is_empty() {
                    "/".to_string()
                } else {
                    format!("/{raw}")
                };
                Finding {
                    code: r.code.clone(),
                    rule_id: r.id.clone(),
                    severity: Severity::Error,
                    path,
                    message: format!("{e}"),
                    expected: None,
                    actual: None,
                    spec_ref: r.spec_ref.clone(),
                    meta: Some(serde_json::json!({ "keyword": "jsonschema" })),
                }
            })
            .collect(),
    };
    findings
}
